File: backend/app/modules/management_governance/assessment.py
"""Governance Assessment — ports frontend assessment.ts."""

from __future__ import annotations

from typing import Any

from app.modules.management_governance.compute import compute_management_governance_model
from app.modules.management_governance.constants import (
    GOVERNANCE_ASSESSMENT_GROUP_LABELS,
    GOVERNANCE_ASSESSMENT_GROUPS,
    GOVERNANCE_CRITERION_STATE_LABELS,
    MANAGEMENT_GOVERNANCE_CONFIRMATION_FIELDS,
)
from app.modules.management_governance.directors import compute_directorship_counts
from app.modules.management_governance.progress import calculate_management_governance_progress

_STATE_PRIORITY = [
    "potential_concern",
    "missing_information",
    "pending_appointment",
    "pending_board_approval",
    "pending_shareholder_approval",
    "pending_linked_workstream",
    "pending_professional_confirmation",
    "appears_ready",
    "not_applicable",
]


def _worst_state(states: list[str]) -> str:
    for state in _STATE_PRIORITY:
        if state in states:
            return state
    return "missing_information"


def _derive_result(criteria: list[dict[str, Any]]) -> dict[str, str]:
    has_concern = any(c["state"] == "potential_concern" for c in criteria)
    has_pending_appointment = any(c["state"] == "pending_appointment" for c in criteria)
    has_professional = any(
        c["state"]
        in (
            "pending_professional_confirmation",
            "pending_board_approval",
            "pending_shareholder_approval",
        )
        for c in criteria
    )
    missing_count = sum(1 for c in criteria if c["state"] == "missing_information")

    if has_concern:
        return {
            "result": "potential_concerns_identified",
            "resultLabel": "Potential concerns identified",
            "summary": (
                "One or more governance checks show a potential concern that needs review "
                "before filing readiness can improve."
            ),
        }
    if has_pending_appointment:
        return {
            "result": "pending_appointments",
            "resultLabel": "Pending appointments",
            "summary": "Some Board or management appointments are still proposed or pending.",
        }
    if has_professional:
        return {
            "result": "professional_confirmation_required",
            "resultLabel": "Professional confirmation required",
            "summary": (
                "Some items still need board, shareholder or professional confirmation."
            ),
        }
    if missing_count > len(criteria) / 2:
        return {
            "result": "insufficient_information",
            "resultLabel": "Disclosure readiness in progress",
            "summary": (
                "Much of the management and governance record is still blank or unanswered."
            ),
        }
    return {
        "result": "readiness_in_progress",
        "resultLabel": "Readiness in progress",
        "summary": (
            "Entered information is largely captured; remaining gaps are noted below."
        ),
    }


def build_governance_assessment(
    payload: dict[str, Any],
    model: dict[str, Any],
    progress: dict[str, Any],
    linked_references: dict[str, Any],
) -> dict[str, Any]:
    criteria: list[dict[str, Any]] = []
    applicability = model.get("applicability") or {}
    board_counts = model.get("boardCounts") or {}
    directors_section = payload.get("directorsProfilesAppointmentsAndEligibility") or {}
    directors = [
        d for d in (directors_section.get("directors") or []) if isinstance(d, dict)
    ]

    criteria.append(
        {
            "id": "minimum-board-size",
            "group": "board_composition",
            "label": "Minimum Board size",
            "state": (
                "appears_ready"
                if board_counts.get("current", 0) >= applicability.get("minimumBoardSize", 0)
                else (
                    "potential_concern"
                    if board_counts.get("current", 0) > 0
                    else "missing_information"
                )
            ),
            "reason": (
                f"Current directors: {board_counts.get('current', 0)}; applicable minimum: "
                f"{applicability.get('minimumBoardSize', 0)} "
                f"({applicability.get('listingSegment', 'unknown')} segment)."
            ),
        }
    )

    criteria.append(
        {
            "id": "woman-director-readiness",
            "group": "board_composition",
            "label": "Woman director readiness",
            "state": (
                "not_applicable"
                if not applicability.get("requiresWomanDirector")
                else (
                    "appears_ready"
                    if board_counts.get("women", 0) >= 1
                    else "pending_appointment"
                )
            ),
            "reason": (
                f"{board_counts.get('women', 0)} woman director(s) among current Board."
                if board_counts.get("women", 0) >= 1
                else "No woman director recorded among current Board members."
            ),
        }
    )

    criteria.append(
        {
            "id": "independent-director-readiness",
            "group": "board_composition",
            "label": "Independent director readiness",
            "state": (
                "not_applicable"
                if not applicability.get("requiresIndependentDirectors")
                else (
                    "appears_ready"
                    if board_counts.get("independent", 0)
                    >= applicability.get("minimumIndependentDirectors", 0)
                    else (
                        "pending_appointment"
                        if board_counts.get("independent", 0) > 0
                        else "missing_information"
                    )
                )
            ),
            "reason": (
                f"{board_counts.get('independent', 0)} independent director(s); applicable "
                f"minimum {applicability.get('minimumIndependentDirectors', 0)}."
                if applicability.get("requiresIndependentDirectors")
                else "Independent director count not assumed mandatory for SME listing segment."
            ),
        }
    )

    criteria.append(
        {
            "id": "resident-director-readiness",
            "group": "board_composition",
            "label": "Resident director readiness",
            "state": (
                "not_applicable"
                if not applicability.get("requiresResidentDirector")
                else (
                    "appears_ready"
                    if board_counts.get("resident", 0) >= 1
                    else "missing_information"
                )
            ),
            "reason": (
                f"{board_counts.get('resident', 0)} resident director(s) recorded."
                if board_counts.get("resident", 0) >= 1
                else "Resident director not identified from country of residence."
            ),
        }
    )

    criteria.append(
        {
            "id": "board-vacancies",
            "group": "board_composition",
            "label": "Board vacancies",
            "state": (
                "appears_ready"
                if model.get("vacantSeats", 0) == 0 and board_counts.get("current", 0) > 0
                else (
                    "potential_concern"
                    if model.get("vacantSeats", 0) > 0
                    else "missing_information"
                )
            ),
            "reason": (
                f"{model.get('vacantSeats', 0)} vacant seat(s) recorded."
                if model.get("vacantSeats", 0) > 0
                else "No vacant seats recorded."
            ),
        }
    )

    criteria.append(
        {
            "id": "proposed-vs-current-distinction",
            "group": "board_composition",
            "label": "Current versus proposed appointments distinguished",
            "state": (
                "appears_ready"
                if (
                    any(
                        str(d.get("appointmentStatus") or "").startswith("proposed")
                        for d in directors
                    )
                    and any(d.get("appointmentStatus") == "current" for d in directors)
                )
                or len(directors) > 0
                else "missing_information"
            ),
            "reason": "Directors carry explicit current/proposed appointment status.",
        }
    )

    for director in directors:
        if not str(director.get("din") or "").strip():
            criteria.append(
                {
                    "id": f"din-{director.get('id')}",
                    "group": "director_eligibility",
                    "label": f"DIN — {director.get('fullLegalName') or director.get('id')}",
                    "state": "missing_information",
                    "reason": "Director Identification Number not recorded.",
                }
            )

        directorship_counts = compute_directorship_counts(director)
        eligibility = director.get("eligibility") or {}
        if (
            directorship_counts["totalCurrent"] >= 7
            or eligibility.get("directorshipLimitConcern") == "yes"
        ):
            criteria.append(
                {
                    "id": f"directorship-limit-{director.get('id')}",
                    "group": "director_eligibility",
                    "label": (
                        f"Directorship count — {director.get('fullLegalName') or director.get('id')}"
                    ),
                    "state": "potential_concern",
                    "reason": (
                        f"{directorship_counts['totalCurrent']} current directorship(s) recorded; "
                        "threshold review suggested."
                    ),
                }
            )

        if eligibility.get("sebiDebarment") == "yes" or eligibility.get(
            "stockExchangeDebarment"
        ) == "yes":
            criteria.append(
                {
                    "id": f"debarment-{director.get('id')}",
                    "group": "director_eligibility",
                    "label": (
                        f"Debarment declaration — "
                        f"{director.get('fullLegalName') or director.get('id')}"
                    ),
                    "state": "potential_concern",
                    "reason": (
                        "A debarment or restraint declaration is marked yes — "
                        "explanation required."
                    ),
                }
            )

        if (
            str(director.get("appointmentStatus") or "").startswith("proposed")
            and not str(director.get("boardApprovalDate") or "").strip()
        ):
            criteria.append(
                {
                    "id": f"pending-board-{director.get('id')}",
                    "group": "director_eligibility",
                    "label": (
                        f"Board approval — {director.get('fullLegalName') or director.get('id')}"
                    ),
                    "state": "pending_board_approval",
                    "reason": "Proposed appointment without board approval date recorded.",
                }
            )

    kmp_section = payload.get("kmpSeniorManagementAndOrganisationStructure") or {}
    readiness = kmp_section.get("kmpRoleReadiness") or {}
    for role_key, role_label in (
        ("cfo", "CFO"),
        ("companySecretary", "Company Secretary"),
        ("complianceOfficer", "Compliance Officer"),
    ):
        status = readiness.get(role_key) or ""
        if status == "completed":
            state = "appears_ready"
        elif status == "professional_confirmation_required":
            state = "pending_professional_confirmation"
        elif status == "not_applicable":
            state = "not_applicable"
        elif status == "in_progress":
            state = "pending_appointment"
        else:
            state = "missing_information"
        criteria.append(
            {
                "id": f"kmp-{role_key}",
                "group": "management_coverage",
                "label": f"{role_label} coverage",
                "state": state,
                "reason": (
                    f"Readiness status: {status.replace('_', ' ')}."
                    if status
                    else "Readiness status not recorded."
                ),
            }
        )

    continuity = model.get("continuity") or {}
    if continuity.get("criticalRoleVacancies", 0) > 0:
        criteria.append(
            {
                "id": "critical-vacancies",
                "group": "management_coverage",
                "label": "Critical role vacancies",
                "state": "potential_concern",
                "reason": (
                    f"{continuity.get('criticalRoleVacancies', 0)} critical role "
                    "vacancy/vacancies recorded."
                ),
            }
        )

    for item in model.get("committeeReadiness") or []:
        if not isinstance(item, dict) or not item.get("required"):
            continue
        status = item.get("status")
        if status == "ready":
            state = "appears_ready"
        elif status == "pending":
            state = "pending_appointment"
        elif status == "not_applicable":
            state = "not_applicable"
        else:
            state = "missing_information"
        criteria.append(
            {
                "id": f"committee-{item.get('committeeType')}",
                "group": "board_committees",
                "label": str(item.get("committeeType") or "").replace("-", " "),
                "state": state,
                "reason": item.get("message") or "",
            }
        )

    capital_ownership = linked_references.get("capitalOwnership") or {}
    if not capital_ownership.get("available"):
        criteria.append(
            {
                "id": "capital-ownership-link",
                "group": "remuneration_and_interests",
                "label": "Capital & Ownership shareholding link",
                "state": "pending_linked_workstream",
                "reason": (
                    "Director/KMP shareholding cross-check awaits Capital & Ownership (M2 wiring)."
                ),
            }
        )

    financials_kpis = linked_references.get("financialsKpis") or {}
    if not financials_kpis.get("available"):
        criteria.append(
            {
                "id": "financials-kpis-link",
                "group": "remuneration_and_interests",
                "label": "Financials & KPIs RPT references",
                "state": "pending_linked_workstream",
                "reason": "RPT amount references await Financials & KPIs (M2 wiring).",
            }
        )

    if continuity.get("repeatCfoChanges", 0) > 0 or continuity.get(
        "repeatCompanySecretaryChanges", 0
    ) > 0:
        criteria.append(
            {
                "id": "repeat-kmp-changes",
                "group": "management_continuity",
                "label": "Repeat CFO / Company Secretary changes",
                "state": "potential_concern",
                "reason": (
                    "Multiple CFO or Company Secretary changes in the last three years — "
                    "explanation suggested."
                ),
            }
        )

    policies_section = payload.get("governancePoliciesRptOversightAndConfirmations") or {}
    confirmations = policies_section.get("confirmations") or {}
    unanswered_confirmations = sum(
        1 for key, _ in MANAGEMENT_GOVERNANCE_CONFIRMATION_FIELDS if not confirmations.get(key)
    )

    for key, label in MANAGEMENT_GOVERNANCE_CONFIRMATION_FIELDS:
        confirmed = bool(confirmations.get(key))
        criteria.append(
            {
                "id": f"confirmation-{key}",
                "group": "governance_processes",
                "label": label,
                "state": "appears_ready" if confirmed else "missing_information",
                "reason": "Confirmed." if confirmed else "Not confirmed yet.",
            }
        )

    board_section = payload.get("boardStructureAndIpoGovernanceReadiness") or {}
    ipo_committee = board_section.get("ipoCommittee") or {}
    constituted = ipo_committee.get("constituted") or ""
    criteria.append(
        {
            "id": "ipo-committee",
            "group": "ipo_specific_governance",
            "label": "IPO Committee constituted",
            "state": (
                "appears_ready"
                if constituted == "yes"
                else (
                    "pending_professional_confirmation"
                    if constituted == "not_sure"
                    else "missing_information"
                )
            ),
            "reason": (
                "IPO Committee marked as constituted."
                if constituted == "yes"
                else "IPO Committee constitution not confirmed."
            ),
        }
    )

    price_band = board_section.get("independentDirectorPriceBandProcess") or {}
    criteria.append(
        {
            "id": "price-band-process",
            "group": "ipo_specific_governance",
            "label": "Independent-director price-band process",
            "state": (
                "not_applicable"
                if price_band.get("requiredApplicabilityStatus") == "not-applicable"
                else (
                    "appears_ready"
                    if price_band.get("committeeConstituted") == "yes"
                    else (
                        "pending_professional_confirmation"
                        if price_band.get("requiredApplicabilityStatus")
                        == "professional-confirmation-required"
                        else "missing_information"
                    )
                )
            ),
            "reason": "Price-band committee process captured for IPO readiness review.",
        }
    )

    groups: list[dict[str, Any]] = []
    for group in GOVERNANCE_ASSESSMENT_GROUPS:
        group_criteria = [c for c in criteria if c["group"] == group]
        groups.append(
            {
                "group": group,
                "label": GOVERNANCE_ASSESSMENT_GROUP_LABELS[group],
                "headlineState": _worst_state([c["state"] for c in group_criteria]),
                "criteria": group_criteria,
            }
        )

    counts = {
        "appearsReady": sum(1 for c in criteria if c["state"] == "appears_ready"),
        "potentialConcern": sum(1 for c in criteria if c["state"] == "potential_concern"),
        "missingInformation": sum(1 for c in criteria if c["state"] == "missing_information"),
        "pendingAppointment": sum(1 for c in criteria if c["state"] == "pending_appointment"),
        "pendingBoardApproval": sum(
            1 for c in criteria if c["state"] == "pending_board_approval"
        ),
        "pendingShareholderApproval": sum(
            1 for c in criteria if c["state"] == "pending_shareholder_approval"
        ),
        "pendingLinkedWorkstream": sum(
            1 for c in criteria if c["state"] == "pending_linked_workstream"
        ),
        "pendingProfessionalConfirmation": sum(
            1 for c in criteria if c["state"] == "pending_professional_confirmation"
        ),
        "notApplicable": sum(1 for c in criteria if c["state"] == "not_applicable"),
    }

    result_info = _derive_result(criteria)

    return {
        **result_info,
        "criteria": criteria,
        "groups": groups,
        "counts": counts,
        "metrics": {
            "boardSize": model.get("boardSize", 0),
            "sectionsComplete": progress.get("sectionsComplete", 0),
            "unansweredConfirmations": unanswered_confirmations,
            "pendingAppointments": model.get("pendingAppointments", 0),
            "potentialConcerns": counts["potentialConcern"],
        },
    }


def assess_management_governance(
    payload: dict[str, Any],
    linked_references: dict[str, Any] | None = None,
) -> dict[str, Any]:
    linked = linked_references or {}
    progress = calculate_management_governance_progress(payload)
    model = compute_management_governance_model(payload, linked)
    return build_governance_assessment(payload, model, progress, linked)


# Exported for tests / documentation parity with frontend labels.
GOVERNANCE_CRITERION_STATE_LABELS_EXPORT = GOVERNANCE_CRITERION_STATE_LABELS

File: frontend/lib/management-governance/assessment.ts
/**
 * Deterministic Governance Assessment for Management & Governance (M1, frontend-only).
 *
 * Readiness-focused: never returns compliant/non-compliant or strong/weak management scores.
 */

import {
  computeManagementGovernanceModel,
  type ManagementGovernanceModel,
} from '@/lib/management-governance/compute';
import { computeDirectorshipCounts } from '@/lib/management-governance/directors';
import { MANAGEMENT_GOVERNANCE_CONFIRMATION_FIELDS } from '@/lib/management-governance/options';
import { calculateManagementGovernanceProgress } from '@/lib/management-governance/progress';
import {
  createEmptyLinkedWorkstreamReferences,
  type LinkedWorkstreamReferences,
  type ManagementGovernanceProgress,
} from '@/lib/management-governance/types';
import type { ManagementGovernancePayload } from '@/lib/schemas/management-governance';

export const GOVERNANCE_CRITERION_STATES = [
  'appears_ready',
  'potential_concern',
  'missing_information',
  'pending_appointment',
  'pending_board_approval',
  'pending_shareholder_approval',
  'pending_linked_workstream',
  'pending_professional_confirmation',
  'not_applicable',
] as const;

export type GovernanceCriterionState = (typeof GOVERNANCE_CRITERION_STATES)[number];

export const GOVERNANCE_ASSESSMENT_GROUPS = [
  'board_composition',
  'director_eligibility',
  'management_coverage',
  'board_committees',
  'remuneration_and_interests',
  'management_continuity',
  'governance_processes',
  'ipo_specific_governance',
] as const;

export type GovernanceAssessmentGroup = (typeof GOVERNANCE_ASSESSMENT_GROUPS)[number];

export const GOVERNANCE_CRITERION_STATE_LABELS: Record<GovernanceCriterionState, string> = {
  appears_ready: 'Appears ready',
  potential_concern: 'Potential concern',
  missing_information: 'Missing information',
  pending_appointment: 'Pending appointment',
  pending_board_approval: 'Pending board approval',
  pending_shareholder_approval: 'Pending shareholder approval',
  pending_linked_workstream: 'Pending linked workstream',
  pending_professional_confirmation: 'Pending professional confirmation',
  not_applicable: 'Not applicable',
};

export const GOVERNANCE_ASSESSMENT_GROUP_LABELS: Record<GovernanceAssessmentGroup, string> = {
  board_composition: 'Board composition',
  director_eligibility: 'Director eligibility',
  management_coverage: 'Management coverage',
  board_committees: 'Board committees',
  remuneration_and_interests: 'Remuneration and interests',
  management_continuity: 'Management continuity',
  governance_processes: 'Governance processes',
  ipo_specific_governance: 'IPO-specific governance',
};

export const GOVERNANCE_ASSESSMENT_RESULT_STATES = [
  'insufficient_information',
  'readiness_in_progress',
  'potential_concerns_identified',
  'professional_confirmation_required',
  'pending_appointments',
] as const;

export type GovernanceAssessmentResultState = (typeof GOVERNANCE_ASSESSMENT_RESULT_STATES)[number];

export type GovernanceAssessmentCriterion = {
  id: string;
  group: GovernanceAssessmentGroup;
  label: string;
  state: GovernanceCriterionState;
  reason: string;
};

export type GovernanceAssessmentGroupResult = {
  group: GovernanceAssessmentGroup;
  label: string;
  headlineState: GovernanceCriterionState;
  criteria: GovernanceAssessmentCriterion[];
};

export type GovernanceAssessmentResponse = {
  result: GovernanceAssessmentResultState;
  resultLabel: string;
  summary: string;
  criteria: GovernanceAssessmentCriterion[];
  groups: GovernanceAssessmentGroupResult[];
  counts: Record<
    | 'appearsReady'
    | 'potentialConcern'
    | 'missingInformation'
    | 'pendingAppointment'
    | 'pendingBoardApproval'
    | 'pendingShareholderApproval'
    | 'pendingLinkedWorkstream'
    | 'pendingProfessionalConfirmation'
    | 'notApplicable',
    number
  >;
  metrics: {
    boardSize: number;
    sectionsComplete: number;
    unansweredConfirmations: number;
    pendingAppointments: number;
    potentialConcerns: number;
  };
};

function worstState(states: GovernanceCriterionState[]): GovernanceCriterionState {
  const priority: GovernanceCriterionState[] = [
    'potential_concern',
    'missing_information',
    'pending_appointment',
    'pending_board_approval',
    'pending_shareholder_approval',
    'pending_linked_workstream',
    'pending_professional_confirmation',
    'appears_ready',
    'not_applicable',
  ];
  for (const state of priority) {
    if (states.includes(state)) return state;
  }
  return 'missing_information';
}

function deriveResult(criteria: GovernanceAssessmentCriterion[]): {
  result: GovernanceAssessmentResultState;
  resultLabel: string;
  summary: string;
} {
  const hasConcern = criteria.some((c) => c.state === 'potential_concern');
  const hasPendingAppointment = criteria.some((c) => c.state === 'pending_appointment');
  const hasProfessional = criteria.some(
    (c) =>
      c.state === 'pending_professional_confirmation' ||
      c.state === 'pending_board_approval' ||
      c.state === 'pending_shareholder_approval',
  );
  const missingCount = criteria.filter((c) => c.state === 'missing_information').length;

  if (hasConcern) {
    return {
      result: 'potential_concerns_identified',
      resultLabel: 'Potential concerns identified',
      summary:
        'One or more governance checks show a potential concern that needs review before filing readiness can improve.',
    };
  }
  if (hasPendingAppointment) {
    return {
      result: 'pending_appointments',
      resultLabel: 'Pending appointments',
      summary: 'Some Board or management appointments are still proposed or pending.',
    };
  }
  if (hasProfessional) {
    return {
      result: 'professional_confirmation_required',
      resultLabel: 'Professional confirmation required',
      summary: 'Some items still need board, shareholder or professional confirmation.',
    };
  }
  if (missingCount > criteria.length / 2) {
    return {
      result: 'insufficient_information',
      resultLabel: 'Disclosure readiness in progress',
      summary: 'Much of the management and governance record is still blank or unanswered.',
    };
  }
  return {
    result: 'readiness_in_progress',
    resultLabel: 'Readiness in progress',
    summary: 'Entered information is largely captured; remaining gaps are noted below.',
  };
}

export function buildGovernanceAssessment(
  payload: ManagementGovernancePayload,
  model: ManagementGovernanceModel,
  progress: ManagementGovernanceProgress,
  linkedReferences: LinkedWorkstreamReferences,
): GovernanceAssessmentResponse {
  const criteria: GovernanceAssessmentCriterion[] = [];
  const { applicability, boardCounts } = model;
  const directors = payload.directorsProfilesAppointmentsAndEligibility.directors;

  criteria.push({
    id: 'minimum-board-size',
    group: 'board_composition',
    label: 'Minimum Board size',
    state:
      boardCounts.current >= applicability.minimumBoardSize
        ? 'appears_ready'
        : boardCounts.current > 0
          ? 'potential_concern'
          : 'missing_information',
    reason: `Current directors: ${boardCounts.current}; applicable minimum: ${applicability.minimumBoardSize} (${applicability.listingSegment} segment).`,
  });

  criteria.push({
    id: 'woman-director-readiness',
    group: 'board_composition',
    label: 'Woman director readiness',
    state: !applicability.requiresWomanDirector
      ? 'not_applicable'
      : boardCounts.women >= 1
        ? 'appears_ready'
        : 'pending_appointment',
    reason:
      boardCounts.women >= 1
        ? `${boardCounts.women} woman director(s) among current Board.`
        : 'No woman director recorded among current Board members.',
  });

  criteria.push({
    id: 'independent-director-readiness',
    group: 'board_composition',
    label: 'Independent director readiness',
    state: !applicability.requiresIndependentDirectors
      ? 'not_applicable'
      : boardCounts.independent >= applicability.minimumIndependentDirectors
        ? 'appears_ready'
        : boardCounts.independent > 0
          ? 'pending_appointment'
          : 'missing_information',
    reason: applicability.requiresIndependentDirectors
      ? `${boardCounts.independent} independent director(s); applicable minimum ${applicability.minimumIndependentDirectors}.`
      : 'Independent director count not assumed mandatory for SME listing segment.',
  });

  criteria.push({
    id: 'resident-director-readiness',
    group: 'board_composition',
    label: 'Resident director readiness',
    state: !applicability.requiresResidentDirector
      ? 'not_applicable'
      : boardCounts.resident >= 1
        ? 'appears_ready'
        : 'missing_information',
    reason:
      boardCounts.resident >= 1
        ? `${boardCounts.resident} resident director(s) recorded.`
        : 'Resident director not identified from country of residence.',
  });

  criteria.push({
    id: 'board-vacancies',
    group: 'board_composition',
    label: 'Board vacancies',
    state:
      model.vacantSeats === 0 && boardCounts.current > 0
        ? 'appears_ready'
        : model.vacantSeats > 0
          ? 'potential_concern'
          : 'missing_information',
    reason:
      model.vacantSeats > 0
        ? `${model.vacantSeats} vacant seat(s) recorded.`
        : 'No vacant seats recorded.',
  });

  criteria.push({
    id: 'proposed-vs-current-distinction',
    group: 'board_composition',
    label: 'Current versus proposed appointments distinguished',
    state:
      directors.some((d) => d.appointmentStatus.startsWith('proposed')) &&
      directors.some((d) => d.appointmentStatus === 'current')
        ? 'appears_ready'
        : directors.length > 0
          ? 'appears_ready'
          : 'missing_information',
    reason: 'Directors carry explicit current/proposed appointment status.',
  });

  for (const director of directors) {
    if (!director.din.trim()) {
      criteria.push({
        id: `din-${director.id}`,
        group: 'director_eligibility',
        label: `DIN — ${director.fullLegalName || director.id}`,
        state: 'missing_information',
        reason: 'Director Identification Number not recorded.',
      });
    }

    const directorshipCounts = computeDirectorshipCounts(director);
    if (directorshipCounts.totalCurrent >= 7 || director.eligibility.directorshipLimitConcern === 'yes') {
      criteria.push({
        id: `directorship-limit-${director.id}`,
        group: 'director_eligibility',
        label: `Directorship count — ${director.fullLegalName || director.id}`,
        state: 'potential_concern',
        reason: `${directorshipCounts.totalCurrent} current directorship(s) recorded; threshold review suggested.`,
      });
    }

    if (director.eligibility.sebiDebarment === 'yes' || director.eligibility.stockExchangeDebarment === 'yes') {
      criteria.push({
        id: `debarment-${director.id}`,
        group: 'director_eligibility',
        label: `Debarment declaration — ${director.fullLegalName || director.id}`,
        state: 'potential_concern',
        reason: 'A debarment or restraint declaration is marked yes — explanation required.',
      });
    }

    if (director.appointmentStatus.startsWith('proposed') && !director.boardApprovalDate.trim()) {
      criteria.push({
        id: `pending-board-${director.id}`,
        group: 'director_eligibility',
        label: `Board approval — ${director.fullLegalName || director.id}`,
        state: 'pending_board_approval',
        reason: 'Proposed appointment without board approval date recorded.',
      });
    }
  }

  const kmp = payload.kmpSeniorManagementAndOrganisationStructure;
  for (const role of [
    { key: 'cfo', label: 'CFO' },
    { key: 'companySecretary', label: 'Company Secretary' },
    { key: 'complianceOfficer', label: 'Compliance Officer' },
  ] as const) {
    const status = kmp.kmpRoleReadiness[role.key];
    criteria.push({
      id: `kmp-${role.key}`,
      group: 'management_coverage',
      label: `${role.label} coverage`,
      state:
        status === 'completed'
          ? 'appears_ready'
          : status === 'professional_confirmation_required'
            ? 'pending_professional_confirmation'
            : status === 'not_applicable'
              ? 'not_applicable'
              : status === 'in_progress'
                ? 'pending_appointment'
                : 'missing_information',
      reason: status ? `Readiness status: ${status.replace(/_/g, ' ')}.` : 'Readiness status not recorded.',
    });
  }

  if (model.continuity.criticalRoleVacancies > 0) {
    criteria.push({
      id: 'critical-vacancies',
      group: 'management_coverage',
      label: 'Critical role vacancies',
      state: 'potential_concern',
      reason: `${model.continuity.criticalRoleVacancies} critical role vacancy/vacancies recorded.`,
    });
  }

  for (const item of model.committeeReadiness.filter((c) => c.required)) {
    criteria.push({
      id: `committee-${item.committeeType}`,
      group: 'board_committees',
      label: item.committeeType.replace(/-/g, ' '),
      state:
        item.status === 'ready'
          ? 'appears_ready'
          : item.status === 'pending'
            ? 'pending_appointment'
            : item.status === 'not_applicable'
              ? 'not_applicable'
              : 'missing_information',
      reason: item.message,
    });
  }

  if (!linkedReferences.capitalOwnership.available) {
    criteria.push({
      id: 'capital-ownership-link',
      group: 'remuneration_and_interests',
      label: 'Capital & Ownership shareholding link',
      state: 'pending_linked_workstream',
      reason: 'Director/KMP shareholding cross-check awaits Capital & Ownership (M2 wiring).',
    });
  }

  if (!linkedReferences.financialsKpis.available) {
    criteria.push({
      id: 'financials-kpis-link',
      group: 'remuneration_and_interests',
      label: 'Financials & KPIs RPT references',
      state: 'pending_linked_workstream',
      reason: 'RPT amount references await Financials & KPIs (M2 wiring).',
    });
  }

  if (model.continuity.repeatCfoChanges > 0 || model.continuity.repeatCompanySecretaryChanges > 0) {
    criteria.push({
      id: 'repeat-kmp-changes',
      group: 'management_continuity',
      label: 'Repeat CFO / Company Secretary changes',
      state: 'potential_concern',
      reason: 'Multiple CFO or Company Secretary changes in the last three years — explanation suggested.',
    });
  }

  const unansweredConfirmations = MANAGEMENT_GOVERNANCE_CONFIRMATION_FIELDS.filter(
    (field) => !payload.governancePoliciesRptOversightAndConfirmations.confirmations[field.key],
  ).length;

  for (const field of MANAGEMENT_GOVERNANCE_CONFIRMATION_FIELDS) {
    criteria.push({
      id: `confirmation-${field.key}`,
      group: 'governance_processes',
      label: field.label,
      state: payload.governancePoliciesRptOversightAndConfirmations.confirmations[field.key]
        ? 'appears_ready'
        : 'missing_information',
      reason: payload.governancePoliciesRptOversightAndConfirmations.confirmations[field.key]
        ? 'Confirmed.'
        : 'Not confirmed yet.',
    });
  }

  const ipoCommittee = payload.boardStructureAndIpoGovernanceReadiness.ipoCommittee;
  criteria.push({
    id: 'ipo-committee',
    group: 'ipo_specific_governance',
    label: 'IPO Committee constituted',
    state:
      ipoCommittee.constituted === 'yes'
        ? 'appears_ready'
        : ipoCommittee.constituted === 'not_sure'
          ? 'pending_professional_confirmation'
          : ipoCommittee.constituted === 'no'
            ? 'missing_information'
            : 'missing_information',
    reason:
      ipoCommittee.constituted === 'yes'
        ? 'IPO Committee marked as constituted.'
        : 'IPO Committee constitution not confirmed.',
  });

  const priceBand = payload.boardStructureAndIpoGovernanceReadiness.independentDirectorPriceBandProcess;
  criteria.push({
    id: 'price-band-process',
    group: 'ipo_specific_governance',
    label: 'Independent-director price-band process',
    state:
      priceBand.requiredApplicabilityStatus === 'not-applicable'
        ? 'not_applicable'
        : priceBand.committeeConstituted === 'yes'
          ? 'appears_ready'
          : priceBand.requiredApplicabilityStatus === 'professional-confirmation-required'
            ? 'pending_professional_confirmation'
            : 'missing_information',
    reason: 'Price-band committee process captured for IPO readiness review.',
  });

  const groups: GovernanceAssessmentGroupResult[] = GOVERNANCE_ASSESSMENT_GROUPS.map((group) => {
    const groupCriteria = criteria.filter((c) => c.group === group);
    return {
      group,
      label: GOVERNANCE_ASSESSMENT_GROUP_LABELS[group],
      headlineState: worstState(groupCriteria.map((c) => c.state)),
      criteria: groupCriteria,
    };
  });

  const counts = {
    appearsReady: criteria.filter((c) => c.state === 'appears_ready').length,
    potentialConcern: criteria.filter((c) => c.state === 'potential_concern').length,
    missingInformation: criteria.filter((c) => c.state === 'missing_information').length,
    pendingAppointment: criteria.filter((c) => c.state === 'pending_appointment').length,
    pendingBoardApproval: criteria.filter((c) => c.state === 'pending_board_approval').length,
    pendingShareholderApproval: criteria.filter((c) => c.state === 'pending_shareholder_approval').length,
    pendingLinkedWorkstream: criteria.filter((c) => c.state === 'pending_linked_workstream').length,
    pendingProfessionalConfirmation: criteria.filter(
      (c) => c.state === 'pending_professional_confirmation',
    ).length,
    notApplicable: criteria.filter((c) => c.state === 'not_applicable').length,
  };

  const { result, resultLabel, summary } = deriveResult(criteria);

  return {
    result,
    resultLabel,
    summary,
    criteria,
    groups,
    counts,
    metrics: {
      boardSize: model.boardSize,
      sectionsComplete: progress.sectionsComplete,
      unansweredConfirmations,
      pendingAppointments: model.pendingAppointments,
      potentialConcerns: counts.potentialConcern,
    },
  };
}

export function buildGovernanceAssessmentFromPayload(
  payload: ManagementGovernancePayload,
  linkedReferences: LinkedWorkstreamReferences = createEmptyLinkedWorkstreamReferences(),
): GovernanceAssessmentResponse {
  const progress = calculateManagementGovernanceProgress(payload);
  const model = computeManagementGovernanceModel(payload, linkedReferences);
  return buildGovernanceAssessment(payload, model, progress, linkedReferences);
}

/** Alias used by tests and UI barrels. */
export const assessManagementGovernance = buildGovernanceAssessmentFromPayload;

export type GovernanceAssessment = GovernanceAssessmentResponse;
